ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -2,319 +2,374 @@
-    return PrognoserVector();
+    ROS_ERROR_STREAM("OW_POWER_SYSTEM ERROR: " << err.what()
+                     << " while attempting to initialize asynchronous "
+                     << "prognosers!\nEnsure /config/async_prognoser.cfg exists "
+                     << "and is formatted properly before restarting the "
+                     << "simulation.");
+    return;
   }
-  return result;
-}
 
-bool PowerSystemNode::loadCustomFaultPowerProfile(string path, string custom_file)
-{
-  m_custom_power_fault_sequence = loadPowerProfile(path, custom_file);
+  ROS_INFO_STREAM("Power system node running.");
 
-  // Return false if the sequence was not properly initialized.
-  return (m_custom_power_fault_sequence.size() > 0);
-}
-
-bool PowerSystemNode::initPrognoser()
-{
-  // Create a configuration from a file
-  auto prognoser_config_path = ros::package::getPath("ow_power_system") + "/config/prognoser.cfg";
-  ConfigMap prognoser_config(prognoser_config_path);
+  // This rate object is used to sync the cycles up to the provided Hz.
+  // NOTE: If the cycle takes longer than the provided Hz, nothing bad will
+  //       necessarily occur, but the simulation will be out of sync with real
+  //       time.
+  ros::Rate rate(m_loop_rate_hz);
 
-  // Contruct a new prognoser using the prognoser factory. The prognoser
-  // will automatically construct an appropriate model, observer and predictor
-  // based on the values specified in the config.
-  m_prognoser = PrognoserFactory::instance().Create("ModelBasedPrognoser", prognoser_config);
+  // Start the asynchronous spinner, which will call jointStatesCb as the
+  // joint_states topic is published (50Hz).
+  ros::AsyncSpinner spinner(m_spinner_threads);
+  spinner.start();
 
-  if (m_prognoser == nullptr)
+  // Loop through the PrognoserInputHandlers to update their values and send them to
+  // the bus each loop to trigger predictions.
+  // This loop should run at 0.5Hz, publishing predictions every 2 seconds via rostopics.
+  int start_loops = NUM_MODELS;
+  while(ros::ok())
   {
-    return false;
-  }
+    // Set up value modifiers in each model handler for injection later.
+    injectFaults();
 
-  // Initialize the GSAP prognoser
-  auto init_data = composePrognoserData(m_initial_power, m_initial_voltage, m_initial_temperature);
-  m_prognoser->step(init_data);
+    // DEBUG PRINT
+    if (m_timestamp_print_debug)
+    {
+      printTimestamp(m_power_models[0].input_info.timestamp);
+    }
 
-  m_init_time = system_clock::now();
+    bool draw_warning_displayed = false;
 
-  return true;
-}
+    // Cycle the inputs in each model and send the data sets to the
+    // corresponding prognoser via its message bus. The power system is
+    // locked during this, preventing callback functions from altering
+    // related variables mid-loop.
+    m_processing_power_batch = true;
+    for (int i = 0; i < NUM_MODELS; i++)
```

Comment:
Follow-up on my mech power comment computed at 1Hz here and passed through `runOnce` to each model. It can then be published along with the other battery topics.